carve-bead.lic

```diff
@@ -0,0 +1,146 @@
+
+class CarveBead
+
+  def initialize
+    arg_definitions = [
+      [
+        { name: 'debug', regex: /debug/i, optional: true, description: 'Turn on debug logging' },
+        { name: 'primer', regex: /primer/i, optional: true, description: 'Force primer use, regardless of settings' },
+        { name: 'aspect', regex: /^[A-z0-9\s\-\'\.]+$/i, optional: true, description: 'Aspect to use, regardless of settings' }
+      ]
+    ]
+    args = parse_args(arg_definitions)
+
+    @settings = get_settings
+    @debug = args.debug || @settings['carve-bead']['debug']
+    @theurgy_supply_container = @settings.theurgy_supply_container
+    @immortal_aspect = args.aspect || @settings.immortal_aspect
+    @bag = @settings.crafting_container
+    @bag_items = @settings.crafting_items_in_container
+    @belt = @settings.engineering_belt
+    @water_holder = @settings.water_holder
+    @use_primer = args.primer || @settings['carve-bead']['primer'] || false
+    @knife_adjective = @settings['carve-bead']['knife-adjective'] || 'carving'
+    all_theurgy_data = get_data('theurgy')
+    @data = all_theurgy_data[@hometown]
+
+    @training_spells = @settings.crafting_training_spells
+
+    if @debug
+      DRC.message(
+        "carve-bead-debug:\n" +
+        "- immortal_aspect: #{@immortal_aspect}\n" +
+        "- use_primer: #{@use_primer}\n" +
+        "- knife_adjective: #{@knife_adjective}"
+      )
+    end
+
+    main
+  end
+
+  def main
+    DRCI.stow_hands
+    read_primer if @use_primer
+    make_block
+    carve_bead
+  end
+
+  def read_primer
+    DRCI.get_item_safe("#{@immortal_aspect} primer", @theurgy_supply_container)
+    open_primer
+
+    keep_reading = true
+    while keep_reading
+      case DRC.bput("study my primer", "Roundtime:", "You have already studied this page.", "You should start on the first page", "You've too recently been studying how to carve a")
+      when 'You should start on the first page'
+        reset_primer
+        next
+      when 'You\'ve too recently been studying how to carve a'
+        DRC.message('You cannot switch between primers that fast. Try again later.')
+        DRCI.put_away_item?("#{@immortal_aspect} primer", @theurgy_supply_container)
+        exit
+      end
+
+      case DRC.bput("turn my primer", "You flip forward a page", "You flip to the end of your primer")
+      when 'You flip forward a page'
+        keep_reading = true
+      else
+        keep_reading = false
+      end
+    end
+    DRCI.put_away_item?("#{@immortal_aspect} primer", @theurgy_supply_container)
+  end
+
+  def open_primer
+    DRC.bput("open my primer", "You open the cover of", "The primer is already open")
+  end
+
+  def close_primer
+    DRC.bput("close my primer", ".*")
+  end
+
+  def reset_primer
+    close_primer
+    open_primer
+  end
+
+  def make_block
+    return false unless DRCTH.has_holy_water?(@theurgy_supply_container, @water_holder)
```

Comment:
holy water is usually more complicated than that, because it's a container in a container. The common-theurgy method is trying to handle all that (and might rely on the DRCI method).